config/config_koanf_hooks.go

```diff
@@ -0,0 +1,171 @@
+package config
+
+import (
+	"encoding"
+	"fmt"
+	"reflect"
+	"strconv"
+	"time"
+
+	"github.com/miekg/dns"
+	"github.com/mitchellh/mapstructure"
+)
+
+// nolint:gochecknoglobals
+var (
+	koanfHookTypes = []reflect.Type{
+		reflect.TypeOf(Duration(0)),
+		reflect.TypeOf(Upstream{}),
+		reflect.TypeOf(QTypeSet{}),
+	}
+)
+
+const (
+	queryTypeMax = 110
+)
+
+func hasCustomHook(t reflect.Type) bool {
+	for _, kht := range koanfHookTypes {
+		if kht == t {
+			return true
+		}
+	}
+
+	return false
+}
+
+func queryTypeHookFunc() mapstructure.DecodeHookFuncType {
+	return func(
+		f reflect.Type,
+		t reflect.Type,
+		data interface{}) (interface{}, error) {
+		if f.Kind() == reflect.Slice &&
+			t == reflect.TypeOf(QTypeSet{}) {
+			s := reflect.ValueOf(data)
+
+			var qtypes []dns.Type
+
+			for i := 0; i < s.Len(); i++ {
+				qt := fmt.Sprint(s.Index(i))
+
+				for qi := 0; qi <= queryTypeMax; qi++ {
+					q := dns.Type(qi)
+					if qt == q.String() {
+						qtypes = append(qtypes, q)
+
+						break
+					}
+
+					if qi == queryTypeMax {
+						return nil, fmt.Errorf("unknown DNS query type: %s", qt)
+					}
+				}
```

Comment:
Maybe I misunderstood the code, but I think this should be doable with `dns.StringToType` ([docs](https://pkg.go.dev/github.com/miekg/dns#StringToType)) instead of checking all types.  
It's easy to miss in the `dns` docs though so maybe that's why you did it like this.

Reply:
Yes you guessed correct... I missed it🤦‍♂️